### DISCUSSION TITLE
{crew} ecosystem (crewtopia?)

### DISCUSSION THREAD

**Discussion by wlandau** [wlandau/crew#62] [Ideas]
Here are some initial thoughts about packages to extend `crew`:

* [`targets`](https://github.com/ropensci/targets): https://github.com/ropensci/targets/issues/753. This is the very next thing I will work on in this effort.
* `crew.cluster`: a package with [custom launcher plugins](https://wlandau.github.io/crew/articles/launcher_plugins.html) for traditional clusters like SLURM and SGE. I would like to be the primary maintainer of this one as well, but I only have access to SGE, so I would appreciate help testing on e.g. SLURM, PBS, TORQUE, and LSF.
* `crew.functional`: functional programming support much like `furrr` extends `future`. I could build this one as well.
* `crew.aws`: [custom launcher plugins](https://wlandau.github.io/crew/articles/launcher_plugins.html) for AWS compute services like Lambda, Batch, Fargate, and possibly raw EC2s. I intend to make a start on this one after most of the above, but I may need help, and I would not be opposed  to delegating it entirely if someone else is more ready and eager than I am.
* `crew.k8s`: [custom launcher plugin](https://wlandau.github.io/crew/articles/launcher_plugins.html) for Kubernetes. Would prefer to delegate this one.
* `crew.gcp`: [custom launcher plugins](https://wlandau.github.io/crew/articles/launcher_plugins.html) for Google Cloud services like Google Cloud Run. Would prefer to delegate.

**Comment by wlandau** [comment]
Updates:

* I created `crew.cluster` in an internal company repo, and I hope to open-source it soon. (Currently going through the approval process.) It works really well on my company's Sun Grid Engine cluster. I don't have access to any other kind of cluster, so I will be asking around for help developing the plugins for e.g. SLURM.
* ~~I am having second thoughts about `crew.functional`. I am not sure it would be a good fit, unless the user creates a fresh controller for each `map()` and the controller is empty beforehand.~~

**Reply by brendanf** [reply]
I'd be happy to help with SLURM when you're ready.

**Reply by wlandau** [reply]
@brendanf, the SLURM plugin to `crew.cluster` is ready for testing: https://wlandau.github.io/crew.cluster/reference/crew_controller_slurm.html, c.f. https://github.com/wlandau/crew.cluster/issues/1. If you have a chance, would you be willing to try it out? I have a couple quick tests at https://github.com/wlandau/crew.cluster/tree/main/tests/slurm which test basic job submission etc. The code at https://github.com/wlandau/crew.cluster/blob/3e8d0d74685f8f7d9005e65327ecdb0386258261/R/crew_launcher_slurm.R#L245-L282 constructs the script (and there is an `args_terminate()` method which tells the launcher to use `--name` when supplying the job name for deletion).

**Comment by wlandau** [comment]
I have been thinking more about functional `crew` (whether it's `crew.functional` or another name), and I actually do want to go ahead and build it after all. I think it would add a lot of value because it would bring all the `crew` [launcher plugins](https://wlandau.github.io/crew/articles/launcher_plugins.html) into the functional map/reduce-style programming ecosystem, which could be really powerful. And think it would actually be fine to assume the controller is fresh and empty to start.

**Reply by kkmann** [reply]
Hey Will, {crew} looks great! Still struggling with wrapping my head around how it fits in my future(.batchtools)/clustermq  R parallel world but the level of abstraction seems to make switching between different environments a lot easier.

The functional interface a la furrr would help a lot with getting people started. In my experience, once you adopt a tidyverse style, a lot of workflows are either dplyr/furrr based or use targets.

Re name, how about fun.crew? ;)

**Reply by wlandau** [reply]
Thanks Kevin! I hope to have time to tackle functional `crew` sometime around late Q2 / early Q3. Nice suggestion on the name.
